pkg/gadgets/run/tracer/helpers.go

```diff
@@ -56,3 +77,259 @@ func getEventTypeBTF(progContent []byte, metadata *types.GadgetMetadata) (*btf.S
+		Insecure: params.Get(insecureParam).AsBool(),
+	}
+	gadget, err := oci.GetGadgetImage(context.TODO(), args[0], authOpts, params.Get(pullParam).AsString())
+	if err != nil {
+		return nil, fmt.Errorf("getting gadget image: %w", err)
+	}
+
+	ret := &types.GadgetInfo{
+		ProgContent:    gadget.EbpfObject,
+		GadgetMetadata: &types.GadgetMetadata{},
+	}
+
+	spec, err := loadSpec(ret.ProgContent)
+	if err != nil {
+		return nil, err
+	}
+
+	if bytes.Equal(gadget.Metadata, ocispec.DescriptorEmptyJSON.Data) {
+		// metadata is not present. synthesize something on the fly from the spec
+		if err := ret.GadgetMetadata.Populate(spec); err != nil {
+			return nil, err
+		}
+	} else {
+		validate := params.Get(validateMetadataParam).AsBool()
+
+		if err := yaml.Unmarshal(gadget.Metadata, &ret.GadgetMetadata); err != nil {
+			return nil, fmt.Errorf("unmarshaling metadata: %w", err)
+		}
+
+		if err := ret.GadgetMetadata.Validate(spec); err != nil {
+			if !validate {
+				logger.Warnf("gadget metadata is not valid: %v", err)
+			} else {
+				return nil, fmt.Errorf("gadget metadata is not valid: %w", err)
+			}
+		}
+	}
+
+	if err := fillTypeHints(spec, ret.GadgetMetadata.EBPFParams); err != nil {
+		return nil, fmt.Errorf("fill parameters type hints: %w", err)
+	}
+
+	ret.GadgetType, err = getGadgetType(ret.GadgetMetadata)
+	if err != nil {
+		return nil, err
+	}
+
+	ret.Features, err = getGadgetFeatures(spec, ret)
+	if err != nil {
+		return nil, fmt.Errorf("getting gadget features: %w", err)
+	}
+
+	// needs to be done after getGadgetFeatures
+	ret.OperatorsParamsCollection = operators.GetOperatorsForContainerizedGadget(ret).ParamDescCollection()
+
+	return ret, nil
+}
+
+// getGadgetType returns the type of the gadget according to the gadget being run.
+func getGadgetType(gadgetMetadata *types.GadgetMetadata) (gadgets.GadgetType, error) {
+	switch {
+	case len(gadgetMetadata.Tracers) > 0:
+		return gadgets.TypeTrace, nil
+	case len(gadgetMetadata.Snapshotters) > 0:
+		return gadgets.TypeOneShot, nil
+	default:
+		return gadgets.TypeUnknown, fmt.Errorf("unknown gadget type")
+	}
+}
+
+func getGadgetFeatures(spec *ebpf.CollectionSpec, info *types.GadgetInfo) (types.GadgetFeatures, error) {
+	features := types.GadgetFeatures{}
+
+	eventType, err := getEventTypeBTF(info.ProgContent, info.GadgetMetadata)
+	if err != nil {
+		return types.GadgetFeatures{}, fmt.Errorf("getting value struct: %w", err)
+	}
+
+	for _, member := range eventType.Members {
+		switch member.Type.TypeName() {
+		case types.MntNsIdTypeName:
+			features.HasMountNs = true
+		case types.L3EndpointTypeName, types.L4EndpointTypeName:
+			features.HasEndpoints = true
+		}
+	}
+
+	for _, p := range spec.Programs {
+		if p.Type == ebpf.SocketFilter && strings.HasPrefix(p.SectionName, "socket") {
+			features.IsAttacher = true
+			// TODO: info.Features.HasNetNs = true
+			// However it creates an issue because the netns column is added twice!
```

Comment:
Check if it's still true after https://github.com/inspektor-gadget/inspektor-gadget/pull/2191